cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -54,61 +56,265 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
+	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryURL) != 0 {
+		if len(cfg.Token) != 0 || len(cfg.DiscoveryToken) != 0 || len(cfg.DiscoveryFile) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url is mutually exclusive with --token, --discovery-token, and --discovery-file",
+				),
+			)
+		}
+		if len(cfg.Masters) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url must not specificy --masters",
+				),
+			)
+		}
+		if len(cfg.TLSBootstrapToken) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"--discovery-url must also specificy TLSBootstrapToken",
+				),
+			)
+		}
 	}
 	return allErrs
 }
 
-func ValidateFileDiscovery(c *kubeadm.FileDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateJoinMastersArgs(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	for _, m := range c.Masters {
+		_, _, err := net.SplitHostPort(m)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, nil, err.Error()))
+		}
+	}
 	return allErrs
 }
 
-func ValidateHTTPSDiscovery(c *kubeadm.HTTPSDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateDiscoveryFile(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	if _, err := os.Stat(c.DiscoveryFile); os.IsNotExist(err) {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "file does not exist"))
+	}
+
 	return allErrs
 }
 
-func ValidateTokenDiscovery(c *kubeadm.TokenDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateDiscoveryURL(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if len(c.ID) == 0 || len(c.Secret) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, nil, "token must be specific as <ID>:<Secret>"))
+	u, err := url.Parse(c.DiscoveryURL)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "invalide URL"))
 	}
-	if len(c.Addresses) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, nil, "at least one address is required"))
+	if u.Scheme != "https" {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "must be https"))
 	}
 	return allErrs
 }
 
-func ValidateAuthorizationMode(authzMode string, fldPath *field.Path) field.ErrorList {
-	if !authzmodes.IsValidAuthorizationMode(authzMode) {
-		return field.ErrorList{field.Invalid(fldPath, nil, "invalid authorization mode")}
+func ValidateToken(t string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	id, secret, err := tokenutil.ParseToken(t)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, err.Error()))
 	}
-	return field.ErrorList{}
+
+	if len(id) == 0 || len(secret) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "token must be specific as <ID>:<Secret>"))
```

Comment:
It may be that this is ok at HEAD so you get it when you rebase, but it should be the `.` separator